src/libos/src/process/do_robust_list.rs

```diff
@@ -0,0 +1,121 @@
+use std::ptr::NonNull;
+use std::sync::atomic::{AtomicU32, Ordering};
+
+use crate::prelude::*;
+
+pub fn do_set_robust_list(list_head_ptr: *mut RobustListHead, len: usize) -> Result<()> {
+    debug!(
+        "set_robust_list: list_head_ptr: {:?}, len: {}",
+        list_head_ptr, len
+    );
+    if std::mem::size_of::<RobustListHead>() != len {
+        return_errno!(EINVAL, "unknown size of RobustListHead");
+    }
+    let robust_list = NonNull::new(list_head_ptr);
+    let current = current!();
+    current.set_robust_list(robust_list);
+    Ok(())
+}
+
+pub fn do_get_robust_list(tid: pid_t) -> Result<*mut RobustListHead> {
+    debug!("get_robust_list: tid: {}", tid);
+    let thread = if tid == 0 {
+        current!()
+    } else {
+        super::table::get_thread(tid)?
+    };
+    let robust_list_ptr = thread
+        .robust_list()
+        .map(|robust_list| robust_list.as_ptr())
+        .unwrap_or(std::ptr::null_mut());
+    Ok(robust_list_ptr)
+}
+
+const FUTEX_WAITERS: u32 = 0x8000_0000;
+const FUTEX_OWNER_DIED: u32 = 0x4000_0000;
+const FUTEX_TID_MASK: u32 = 0x3FFF_FFFF;
+const ROBUST_LIST_LIMIT: u16 = 2048;
+
+/// This struct is same with Linux's robust_list
+#[repr(C)]
+struct RobustList {
+    next: *const RobustList,
+}
+
+/// This struct is same with Linux's robust_list_head
+#[repr(C)]
+pub struct RobustListHead {
+    list: RobustList,
+    futex_offset: usize,
+    list_op_pending: *const RobustList,
+}
+
+impl RobustListHead {
+    pub fn wake(&self, tid: pid_t) -> Result<()> {
+        debug!("wake rubust list of the dying thread(tid: {})", tid);
+        let mut list_cnt = 0;
+        let mut entry_ptr = self.list.next;
+        while entry_ptr != &self.list {
```

Comment:
while entry_ptr != &self.list {
        if entry_ptr.is_null() {
            return_errno!(EFAULT, "entry_ptr is null pointer");
        }
        
        if entry_ptr != self.list_op_pending {
            let futex_key_addr = unsafe { Self::futex_key_addr(entry_ptr, self.futex_offset) };
            Self::wake_one_futex(futex_key_addr, tid)?;
        }
        
        let next_entry_ptr = {
            unsafe { (*entry_ptr).next }
        };
        entry_ptr = next_entry_ptr;
        list_cnt += 1;
        // Avoid excessively long or circular lists
        if list_cnt >= ROBUST_LIST_LIMIT {
            break;
        }
 }

Is it the readability is better?